add feature disableAlwaysUseSendBeacon

Orgoth opened the pull request:
### Description:

https://github.com/matomo-org/matomo/issues/19864

> The uBlock extension no longer blocks Matomo calls exclusively via the parameters or name.
EasyPrivacy has now also included the request type and blocks the initial ping.
Would it be possible to make the request type also adjustable?

> It seems the TagManager currently only supports to send alwaysUseSendBeacon, but not the disable method. But my experience with TagManager is limited, so maybe create an issue in the tagmanager repo for supporting the disableAlwaysUseSendBeacon method: https://github.com/matomo-org/tag-manager/issues. They might also be able to provide a workaround till it is supported.

https://github.com/matomo-org/tag-manager/issues/581

Yes it is not really clear for the user, since there is already the entry for "Always use sendBeacon" and adding an option for "Disable Always use sendBeacon" would add confusion to the end user.

But since unfortunately nothing has happened for more than half a year and another user asked for a PR, I have now provided one here.

The changes are already in use in my production environment and I keep patching them with every update of Matomo.

### Review

* [x] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

Commit:
add feature disableAlwaysUseSendBeacon

Commit:
updated testfiles for the new fields

EPNW-Eric commented:
Thanks for the PR!

As a suggestion to
> Yes it is not really clear for the user, since there is already the entry for "Always use sendBeacon" and adding an option for "Disable Always use sendBeacon" would add confusion to the end user.

the same is the case with various other settings like `customCookieTimeOutEnable`, which is only visible if `disableCookies` is unchecked. This is solved by a `$field->condition = '!disableCookies';`:
```php
$this->makeSetting('customCookieTimeOutEnable', false, FieldConfig::TYPE_BOOL, function (FieldConfig $field) {
    $field->title = Piwik::translate('TagManager_MatomoConfigurationMatomoCustomCookieTimeOutEnableTitle');
    $field->description = Piwik::translate('TagManager_MatomoConfigurationMatomoCustomCookieTimeOutEnableDescription');
    $field->condition = '!disableCookies';
}),
```
So making `alwaysUseSendBeacon` mutually exclusive with `disableAlwaysUseSendBeacon` should be possible by adding a `$field->condition` to both settings?

Commit:
corrected order of the rows within the test.xml
alwaysUseSendBeacon must be first, then disableAlwaysUseSendBeacon

Commit:
added condition "$field->condition = '!alwaysUseSendBeacon';"

ghost commented:
> Thanks for the PR!
> 
> As a suggestion to
> 
> > Yes it is not really clear for the user, since there is already the entry for "Always use sendBeacon" and adding an option for "Disable Always use sendBeacon" would add confusion to the end user.
> 
> the same is the case with various other settings like `customCookieTimeOutEnable`, which is only visible if `disableCookies` is unchecked. This is solved by a `$field->condition = '!disableCookies';`:
> 
> ```
> $this->makeSetting('customCookieTimeOutEnable', false, FieldConfig::TYPE_BOOL, function (FieldConfig $field) {
>     $field->title = Piwik::translate('TagManager_MatomoConfigurationMatomoCustomCookieTimeOutEnableTitle');
>     $field->description = Piwik::translate('TagManager_MatomoConfigurationMatomoCustomCookieTimeOutEnableDescription');
>     $field->condition = '!disableCookies';
> }),
> ```
> 
> So making `alwaysUseSendBeacon` mutually exclusive with `disableAlwaysUseSendBeacon` should be possible by adding a `$field->condition` to both settings?

Thank you for the suggestion, I have added the condition.

sgiehl commented:
@Orgoth As we are already preparing the release of Matomo 5, we are not planning to add new features for Matomo 4. It would therefor be needed to target `5.x-dev` branch with your changes

Commit:
Merge branch '5.x-dev' into 4.x-dev-patch

Commit:
added condition "$field->condition = '!disableAlwaysUseSendBeacon';" for alwaysUseSendBeacon

ghost commented:
@sgiehl is changed and branch updated to 5.x-dev

Commit:
updated testcases regarding the conditions

Commit:
updated testcases regarding the conditions

ghost commented:
@EPNW-Eric Have a look at all the commits only to get my small workaround into the main branch.

At the moment the system is complaining, that the image size of the test screenshots differ and I don't know how to fix this case.

https://github.com/matomo-org/tag-manager/actions/runs/5901621311/job/16010065823?pr=676

I hope someone from the core team will review and help fixing these test cases.

@sgiehl @snake14

AltamashShaikh commented on lang/en.json:
```diff
@@ -578,6 +578,8 @@
         "MatomoConfigurationMatomoAlwaysUseSendBeaconTitle": "Always use sendBeacon",
         "MatomoConfigurationMatomoAlwaysUseSendBeaconDescription": "Enables sendBeacon usage instead of a regular ajax request. This means when a user clicks for example on an outlink, the navigation to this page will happen much faster.",
+        "MatomoConfigurationMatomoDisableAlwaysUseSendBeaconTitle": "Disable always use sendBeacon permanently",
+        "MatomoConfigurationMatomoDisableAlwaysUseSendBeaconDescription": "Disable sendBeacon usage instead of a regular ajax request. This means when a user clicks for example on an outlink, the navigation to this page will be slower, but it will not be blocked by an Adblocker.",
```
@Orgoth Some text suggestion 
```suggestion
        "MatomoConfigurationMatomoDisableAlwaysUseSendBeaconDescription": "Choose to stop using sendBeacon and use a regular AJAX request instead. This might make clicking on external links a bit slower for users, but it won't be stopped by ad blockers.",
```

AltamashShaikh reviewed (commented):
@Orgoth Thanks for the Pr, overall the code looks good to me
The failing UI test can be fixed by any one of us when its merged

I have suggested 1 textual change, but its not important

snake14 reviewed (approved):
Nice work @Orgoth :+1: 
My only thought is that I like the description wording change that @AltamashShaikh suggested. It seems a little easier to read and understand.

Commit:
Updated to the suggestion by @AltamashShaikh
- https://github.com/matomo-org/tag-manager/pull/676#discussion_r1298641706

ghost commented:
Thank you @AltamashShaikh @snake14 
I have added the suggestion as commit.

Commit:
Update en.json

Forgotten to change the language file itself :) did only change the testcase
The new text is now in both files.
